### DISCUSSION TITLE
Run Flood and torrent client on different platforms

### DISCUSSION THREAD

**Discussion by xkillermaverick** [jesec/flood#195] [Ideas]
Type: Bug Report

## Your Environment

- Version used:
  - Version (stable release) : 4.4.1 (with NPM)
  - Commit ID (development release) `git --no-pager log -1` : 
- Environment name and version:
  - Node.js version `node --version : v15.8.0
  - npm version `npm --version : v7.5.3
  - Operating system and version: Fedora 33 (Server Edition) Kernel 5.10.13-200.fc33.x86_64
- Torrent client and version: qBittorrent v4.3.3
  - Operating system and version: Windows 10 Version 20H2 Build 19042.804
- Web browser `name and version : Microsoft Edge Version 88.0.705.63

## Possible Solution

Remove the path correction code, Improve the path correction to recognize a windows path "A:/", Make path correction code Toggleable by the user.

## Steps to Reproduce

1. Install flood on Linux
2. Install Torrent Client on Windows
3.  qBittorrent Errors out due to the path looking like /A:/Torrent/Downloads/ when it expects A:/Torrent/Downloads/

**Comment by jesec** [comment]
This is not a supported case. README very clearly notes that you have to run Flood and torrent client on the same machine.

**Comment by xkillermaverick** [comment]
ok but everything works this is the only issue,

**Comment by jesec** [comment]
Flood needs the same filesystem context, which is not achievable when Flood is not on the same platform as torrent client. You may get a minimally working web UI but many features DO break.

**Comment by xkillermaverick** [comment]
i understand that but none the less why do even need to add / in the path if it does not have one? why not add it before the user adds the torrent. actually can you message what features break in flood?

**Comment by jesec** [comment]
Because that’s how Unix path works. This is the expected behavior of path expansion function when executed on a Linux machine. It simply can’t go other way. Otherwise, things can break for people who do have a **correct** setup.

Some features that break:
- Browse folders (filesystem browser)
- Disk usage
- Download contents
- Download .torrent
- Stream contents
- mediainfo
- etc. depends on your torrent client. certain clients have increased reliance on Flood

**Comment by afontenot** [comment]
Note that this used to work fine, with a few broken features, as noted above. I was totally fine with this, honestly, because I don't use any of those features. I think the totally arbitrary check is pretty annoying, and if someone needs help patching it out, send me a ping.

That said, I don't think separating Flood from your client provides real security, for most clients anyway. Getting access to rtorrent's XML-RPC lets you straight up run commands as the user rtorrent is running under, IIRC. (This implies that Flood has access to all your files anyway.)

**Reply by jesec** [reply]
What check? There is no such check. I don't understand what you mean.

**Reply by afontenot** [reply]
Sorry for being unclear, I'm referring to "Connection could not be verified." As I mentioned in my comment on the other discussion, it's rather unclear what it means. I understood it to mean a check for permissions was being performed. If you want a user story, it looks something like this:

1. I originally used the previous version of Flood (before you forked it) with no issues, other than not being able to delete / manage files because I don't give Flood access to the download folder rtorrent is using.

2. I switched to your fork, recreating my config from scratch. After entering the details, instead of a "connection failed" or similar message, I got "Connection could not be verified."

3. I didn't understand what there could be to verify. What exactly is Flood checking *other* than the connection itself? I checked that rtorrent was putting the socket file in a directory that the flood user has read / write access to (it was).

4. Still confused, I checked the README. The very first lines in the troubleshooting section say "Flood needs to have permissions/access to the files" and "Flood only uses the path provided by the torrent client so it needs to have the same filesystem context as the torrent client."

5. "Huh", I said to myself. "I'm definitely not doing that. I guess jesec changed the code in such a way that it forces you to use this particular configuration." As a test, I moved the Flood installation over to the rtorrent user, and ran it there. It immediately worked, so I moved on and assumed that I had been right about this.

Thinking about it, it's possible that rtorrent was creating the socket file with 700 permissions or something, and I just missed it. So maybe I misunderstood the situation, in which case I apologize for the distraction. If that's the case, as I said in the other thread, I think "Connection could not be verified" is a very vague and misleading error message. Better to use "Flood could not connect to rtorrent", which is at least precise and informative if for some reason there aren't any diagnostics to provide.

**Reply by jesec** [reply]
In my opinion, the messages are quite clear. Frankly, unless you are visually impaired, you should see the big and bold "Cannot connect to the client", and I don't think there is any ambiguity with "can't be verified". It fails because it fails. **You** have to deal with that. 

![image](https://user-images.githubusercontent.com/11585141/107917804-7ba9a580-6fa3-11eb-9264-4b039efdc4e4.png)

And, BTW: I don't know how this is related to the OP.

**Reply by afontenot** [reply]
I don't see any need to be rude.

"Could not be verified" does **not** mean the same thing as "cannot connect". In fact the implication of using a circumlocution like "can't be verified" is that you are *deliberately* **not** saying that it couldn't connect. In other words, the takeaway for the alert user is "I can connect, but something else went wrong when I tried to ensure the connection will work / remain stable." And then the user is left to guess what that is. The very fact that we are having this conversation is proof that at least *one* person found the message to be misleading, whether you do or not. (The screenshot you show is not the error I got, by the way.)

Also, it's one thing when you're writing software for your own use, but another thing when you're trying to make something for a general audience (or, as in this case, taking over development of a project for a general audience). "It fails because it fails" would be a bad error message, obviously. If there are obvious foot cannons, you should probably try to avoid those. If there is information you could provide that would enable the user to debug a problem more easily, you should provide it. Not every piece of open source software is for people who make six figs at Google developing targeted advertising. I shouldn't have to read the code or dig through a mountain of documentation to figure out what went wrong when it's as simple as "the permissions on the socket file are too strict".

**Reply by jesec** [reply]
I apologize. To be honest, it is quite frustrating. I think users **should** at least have some skills to figure out their own stuffs. README and other documentations already provide more than enough information. If some users think they are advanced enough to use some weird setups, they shall take up the responsibilities and deal with the troubles (instead of complaining about error messages). 

Unfortunately, from a security perspective, it is unwise to expose too much details about system internal to the frontend. Generally, when such a connection failure happens, you will see logs and traces on the server side instead. And, with a proper setup, such an error shall never happen.

I don't think there is anything I can do. However, if someone else feels there is such a need, I may as well accept a PR.

**Reply by afontenot** [reply]
Thanks, and also, I'm sorry for responding a bit dramatically and generally wasting your time. 

I agree with you about leaving advanced setups for more knowledgeable users. To be clear, I *was* able to fix the issue myself very quickly - my complaints about it here have all been with the goal of highlighting what seems to me to be a confusing error message. I think you're likely to see other users confused by it in the future.

For the record, I was able to replicate the error. It looks like this:

![Screenshot_20210215_003803](https://user-images.githubusercontent.com/3870410/107923690-b0c3e280-6f26-11eb-9039-a31141053658.png)

You might ask why I ignored the "Cannot connect to the client" message. The reason is that that text was *already* in the modal (it's the modal's title) *before* I clicked "Save Settings". I got to this by having a broken configuration after updating from a previous version of Flood. Initially the modal said that it couldn't connect, and then I added the path to the socket, clicked Save, and got the "Connection could not be verified", which popped up in red. So my read of the situation was that the purpose of this modal was to fix my connection, hence the title ("Cannot connect to the client"), and that after entering the path, the popup message in red was the result of attempting to connect. In fact it was probably the *contrast* between the two messages that led me to believe that "verification" referred to something different. Would a smarter person than me have immediately recognized the intended meaning? Maybe.

Sorry if my story is boring, I just find that in development it's often useful to have user stories which indicate how they interacted with parts of a UI.

Not 100% sure I understand what you mean by a security perspective here, I would've assumed no one except logged in users are ever supposed to see the "cannot connect to the client" modal. Once you do the initial setup, Flood is presumably supposed to be secure against additional users gaining access. So my thinking was that Flood could give the user a message like "Flood does not have write permission to the socket" (or even "EACCES -13 /tmp/rtorrent.sock", which is at least searchable).

**Reply by xkillermaverick** [reply]
Hey have you check SELinux if flood does not have permission then yes its something on your end

**Reply by jesec** [reply]
Can someone explain to me how this thread is related to the feature request of "run Flood on Linux and run torrent client on Windows"?

I am hiding this thread since it does not seem to relate to the OP. Things are getting more confusing as time passes.

As for the SELinux, if you think you are advanced to use that, **you** should deal with troubles. Flood does not have that sort of checks.
